apps/passport-server/src/services/generic-issuance/SemaphoreGroupProvider.ts

```diff
@@ -0,0 +1,309 @@
+   * the eventId/productId pairs to match configured Semaphore groups, and
+   * build up membership sets for each group. We then compare that to the
+   * in-memory representation, and then add or remove from the in-memory group
+   * representation as required.
+   */
+  private async updateGroup(
+    groupConfig: SemaphoreGroupConfig,
+    emails: string[]
+  ): Promise<void> {
+    return traced(LOG_NAME, "updateGroup", async (span) => {
+      const consumers =
+        emails.length > 0
+          ? await this.consumerDB.loadByEmails(this.pipelineId, emails)
+          : [];
+
+      const semaphoreIds = consumers.map((consumer) => consumer.commitment);
+      const group = this.groups.get(groupConfig.groupId);
+      if (!group) {
+        throw new Error(
+          `Could not find group ${groupConfig.groupId} on pipeline ${this.pipelineId}`
+        );
+      }
+
+      const { toAdd, toRemove } = this.calculateGroupChanges(
+        group,
+        semaphoreIds
+      );
+
+      if (toAdd.length > 0 || toRemove.length > 0) {
+        for (const newId of toAdd) {
+          group.addMember(newId);
+        }
+
+        for (const deletedId of toRemove) {
+          group.removeMember(group.indexOf(BigInt(deletedId)));
+        }
+
+        await this.semaphoreHistoryDB.add(
+          this.pipelineId,
+          groupConfig.groupId,
+          group.root.toString(),
+          JSON.stringify(serializeSemaphoreGroup(group, groupConfig.name))
+        );
+      }
+      span?.setAttribute("pipeline_id", this.pipelineId);
+      span?.setAttribute("emails_count", emails.length);
+      span?.setAttribute("consumers_count", consumers.length);
+      span?.setAttribute("group_id", groupConfig.groupId);
+      span?.setAttribute("added_count", toAdd.length);
+      span?.setAttribute("removed_count", toRemove.length);
+      span?.setAttribute("total_members_after_update", group.members.length);
+    });
+  }
+
+  /**
+   * Calculates the changes to the group compared to the latest set of
+   * members.
+   */
+  private calculateGroupChanges(
+    group: Group,
+    latestMembers: string[]
+  ): { toAdd: string[]; toRemove: string[] } {
+    const groupMembers = group.members
+      .filter((m) => m !== group.zeroValue)
+      .map((m) => m.toString());
+
+    const groupMemberSet = new Set(groupMembers);
+
+    const toAdd = latestMembers.filter((id) => !groupMemberSet.has(id));
+    const latestMemberSet = new Set(latestMembers);
+    const toRemove = groupMembers.filter(
+      (id) => !latestMemberSet.has(id.toString())
+    );
+
+    return {
+      toAdd,
+      toRemove
+    };
+  }
+
+  /**
+   * Given data on the current set of tickets (specifically event ID, product
+   * ID, and email), match these to configured Semaphore groups.
+   */
+  private groupEmailsFromTicketData(
+    ticketData: SemaphoreGroupTicketInfo[]
+  ): Map<SemaphoreGroupConfig, string[]> {
+    const groups = new Map<SemaphoreGroupConfig, string[]>();
```

Comment:
I think what I really want here is a tuple, rather than a map.

Reply:
It could be a map of group config IDs to the emails that match the configuration's specification for group membership. The problem is that then the caller is left looking up a group config by ID, and I have to do the silly "if the group config referred to by this ID doesn't exist, throw an exception" thing.